packages/go_router/lib/src/builder.dart

```diff
@@ -259,6 +336,12 @@ class RouteBuilder {
       if (pageBuilder != null) {
         page = pageBuilder(context, state);
       }
+    } else if (route is StatefulShellRoute) {
+      final ShellRoutePageBuilder? pageForShell = route.pageBuilder;
+      assert(child != null, 'StatefulShellRoute must contain a child route');
+      if (pageForShell != null) {
+        page = pageForShell(context, state, child!);
+      }
```

Comment:
This code is more or less identical to the else if block below, and can be merged with it.